invokeai/backend/grounded_sam/detection_result.py

```diff
@@ -0,0 +1,41 @@
+from dataclasses import dataclass
+from typing import Any, Optional
+
+import numpy.typing as npt
+
+
+@dataclass
+class BoundingBox:
+    """Bounding box helper class."""
+
+    xmin: int
+    ymin: int
+    xmax: int
+    ymax: int
+
+    def to_box(self) -> list[int]:
+        """Convert to the array notation expected by SAM."""
+        return [self.xmin, self.ymin, self.xmax, self.ymax]
+
+
+@dataclass
+class DetectionResult:
+    """Detection result from Grounding DINO or Grounded SAM."""
+
+    score: float
+    label: str
+    box: BoundingBox
+    mask: Optional[npt.NDArray[Any]] = None
+
+    @classmethod
+    def from_dict(cls, detection_dict: dict[str, Any]):
+        return cls(
+            score=detection_dict["score"],
+            label=detection_dict["label"],
+            box=BoundingBox(
+                xmin=detection_dict["box"]["xmin"],
+                ymin=detection_dict["box"]["ymin"],
+                xmax=detection_dict["box"]["xmax"],
+                ymax=detection_dict["box"]["ymax"],
+            ),
+        )
```

Comment:
Given we are manually implementing the `from_dict` method, these classes may want to be pydantic models:

```suggestion
from typing import Any, Optional

import numpy as np
import numpy.typing as npt
from pydantic import BaseModel, ConfigDict


class BoundingBox(BaseModel):
    """Bounding box helper class."""

    xmin: int
    ymin: int
    xmax: int
    ymax: int

    def to_box(self) -> list[int]:
        """Convert to the array notation expected by SAM."""
        return [self.xmin, self.ymin, self.xmax, self.ymax]


class DetectionResult(BaseModel):
    """Detection result from Grounding DINO or Grounded SAM."""

    score: float
    label: str
    box: BoundingBox
    mask: Optional[npt.NDArray[Any]] = None

    model_config = ConfigDict(
        # Allow arbitrary types for mask, since it will be a numpy array.
        arbitrary_types_allowed=True
    )
```

Then one change in `grounding_dino_pipeline.py` @ L20:
```diff
diff --git a/invokeai/backend/grounded_sam/grounding_dino_pipeline.py b/invokeai/backend/grounded_sam/grounding_dino_pipeline.py
index 1fc92b5e1..97c92f924 100644
--- a/invokeai/backend/grounded_sam/grounding_dino_pipeline.py
+++ b/invokeai/backend/grounded_sam/grounding_dino_pipeline.py
@@ -17,7 +17,7 @@ class GroundingDinoPipeline:
 
     def detect(self, image: Image.Image, candidate_labels: list[str], threshold: float = 0.1) -> list[DetectionResult]:
         results = self._pipeline(image=image, candidate_labels=candidate_labels, threshold=threshold)
-        results = [DetectionResult.from_dict(result) for result in results]
+        results = [DetectionResult.model_validate(result) for result in results]
         return results
 
     def to(self, device: Optional[torch.device] = None, dtype: Optional[torch.dtype] = None) -> "GroundingDinoPipeline":
```

Reply:
Adopted in [33e8604](https://github.com/invoke-ai/InvokeAI/pull/6701/commits/33e8604b5753cadd1b07ef8a6e995552986516f6)